tests/unit/test_solvers/test_solution.py

```diff
@@ -371,115 +370,102 @@ def test_plot(self):
+        )
+        data_load = loadmat(f"{test_stub}.mat")
+        np.testing.assert_array_equal(solution.data["c + d"], data_load["c_plus_d"])
+
+        # to csv
+        with pytest.raises(ValueError, match="only 0D variables can be saved to csv"):
+            solution.save_data(f"{test_stub}.csv", to_format="csv")
+        # only save "c" and "2c"
+        solution.save_data(f"{test_stub}.csv", ["c", "2c"], to_format="csv")
+        csv_str = solution.save_data(variables=["c", "2c"], to_format="csv")
+
+        # check string is the same as the file
+        with open(f"{test_stub}.csv") as f:
+            # need to strip \r chars for windows
+            assert csv_str.replace("\r", "") == f.read()
+
+        # read csv
+        df = pd.read_csv(f"{test_stub}.csv")
+        np.testing.assert_allclose(df["c"], solution.data["c"], rtol=1e-7, atol=1e-6)
+        np.testing.assert_allclose(df["2c"], solution.data["2c"], rtol=1e-7, atol=1e-6)
+
+        # to json
+        solution.save_data(f"{test_stub}.json", to_format="json")
+        json_str = solution.save_data(to_format="json")
+
+        # check string is the same as the file
+        with open(f"{test_stub}.json") as f:
+            # need to strip \r chars for windows
+            assert json_str.replace("\r", "") == f.read()
+
+        # check if string has the right values
+        json_data = json.loads(json_str)
+        np.testing.assert_allclose(
+            json_data["c"], solution.data["c"], rtol=1e-7, atol=1e-6
+        )
+        np.testing.assert_allclose(
+            json_data["d"], solution.data["d"], rtol=1e-7, atol=1e-6
+        )
 
-            # raise error if format is unknown
-            with pytest.raises(
-                ValueError, match="format 'wrong_format' not recognised"
-            ):
-                solution.save_data(f"{test_stub}.csv", to_format="wrong_format")
-
-            # test save whole solution
-            solution.save(f"{test_stub}.pickle")
-            solution_load = pybamm.load(f"{test_stub}.pickle")
-            assert solution.all_models[0].name == solution_load.all_models[0].name
-            np.testing.assert_array_equal(
-                solution["c"].entries, solution_load["c"].entries
-            )
-            np.testing.assert_array_equal(
-                solution["d"].entries, solution_load["d"].entries
-            )
+        # raise error if format is unknown
+        with pytest.raises(ValueError, match="format 'wrong_format' not recognised"):
+            solution.save_data(f"{test_stub}.csv", to_format="wrong_format")
+
+        # test save whole solution
+        solution.save(f"{test_stub}.pickle")
+        solution_load = pybamm.load(f"{test_stub}.pickle")
+        assert solution.all_models[0].name == solution_load.all_models[0].name
+        np.testing.assert_array_equal(solution["c"].entries, solution_load["c"].entries)
+        np.testing.assert_array_equal(solution["d"].entries, solution_load["d"].entries)
```

Comment:
Same here (please feel free to push back at any point if you want to limit the changes in the PR).